README
NAME
    Convert::MRC - CONVERT MRC TO TBX-BASIC

VERSION
    version 0.01

SYNOPSIS
            use strict;
            use warnings;

            my $converter = Convert::MRC->new;
            $converter->input_fh('/path/to/MRC/file.mrc');
            $converter->tbx_fh('/path/to/output/file.tbx');
            $converter->log_fh('/path/to/log/file.log');
            $converter->convert;

DESCRIPTION
  MRC
    The MRC format is fully described in an article by Alan K. Melby which
    appeared in Tradumatica
    <http://www.ttt.org/tbx/AKMtradumaArticle-publishedVersion.pdf>. At an
    approximation, it is a file of tab-separated rows, each consisting of an
    ID, a data category, and a value to be stored for that category in the
    object with the given ID. The file should be sorted on its first column.
    If it is not, the converter may skip rows (if they are at too high a
    level) or end processing early (if the order of A-rows, C-rows, and
    R-rows is broken).

  CONVERSION TO TBX-BASIC
    This translator receives a file or list of files in this format and
    emits TBX-Basic, a standard format for terminology interchange.
    Incorrect or unusable input is skipped, with one exception, and the
    problem is noted in a log file. The outputs generally have the same
    filename as the inputs, and a suffix of .tbx and .warnings, but a number
    may be added to the filename to ensure the output filenames are unique.

    The exception noted is this: If the user documents a party responsible
    for some change in the termbase, but does not state whether that party
    is a person or an organization, the party will be included in the TBX as
    a "respParty". This designation does not conform to the TBX-Basic
    standard and will need to be changed (to "respPerson" or "respOrg")
    before the file will validate. This is one of the circumstances in which
    the converter will output invalid TBX-Basic.

    The other circumstance is that a file might not contain a definition, a
    part of speech, or a context sentence for some term, or might not
    contain a term itself. The converter detects these and warns about them,
    but there is no way it could fix them. It does not detect or warn about
    concepts containing no langSet or langSets containing no term, but these
    are also invalid.

NAME
    Convert::MRC- Perl extension for converting MRC files into TBX-Basic.

METHODS
  "new"
    Creates and returns a new instance of Convert::MRC.

  "tbx_fh"
    Optional argument: string file path or GLOB

    Sets and/or returns the file handle used to print the converted TBX.

  "log_fh"
    Optional argument: string file path or GLOB

    Sets and/or returns the file handle used to log any messages.

  "input_fh"
    Optional argument: string file path or GLOB; '-' means STDIN

    Sets and/or returns the file handle used to read the MRC data from.

  "batch"
    Processes each of the input files, printing the converted TBX file to a
    file with the same name and the suffix ".tbx". Warnings are also printed
    to a file with the same name and the suffix ".log".

  "convert"
    Converts the input MRC data into TBX-Basic:

    * Reading MRC data from "input_fh"

    * Printing TBX-Basic data to "tbx_fh"

    * Logging messages to "log_fh"

SEE ALSO
    * The homepage for this program is located here
      <http://tbxconvert.gevterm.net/mrc2tbx/index.html>. You can use it
      online (one file at a time), and can also view a tutorial about MRC
      files.

    * A more in-depth look at MRC can be found in this article
      <http://www.ttt.org/tbx/AKMtradumaArticle-publishedVersion.pdf>.

    * General TBX iformation can be found here <http://www.ttt.org/tbx>.

AUTHOR
    Nathan Rasmussen, Nathan Glenn <garfieldnate@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Alan K. Melby.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changes file for Convert::MRC


{{$NEXT}}

4.03      2013-10-28 22:26:23 UTC
    Updated repository information
    Added annotations to quiet perlcritic
4.02      2013-03-26 21:59:06 UTC
	Added JSON and YAML files, with Git info

4.01      2013-03-26 21:48:33 UTC
	Minor fix to pod formatting

4.0       2013-03-26 21:21:59 UTC
	Initial CPAN release includes extensive updates from version 3.3:
		- extensive test suite
		- object orientation
		- usage of real logger, instead of simple printing
		- mrc2tbx script
		- minor bug fixes

t/TestMRCConverter.pm
package t::TestMRCConverter;
use Test::Base -Base;
use Convert::MRC;

package t::TestMRCConverter::Filter;
use Test::Base::Filter -base;
use Convert::MRC;
use Data::Dumper;

my $converter = Convert::MRC->new();

sub print_self {
	print Dumper $converter;
	return @_;
}

sub convert {
	my ($mrc) = @_;

	#read MRC from input string
	open my $mrc_handle, '<', \$mrc;

	#send output to strings
	my ( $tbx, $log );
	open my $tbx_handle, '>', \$tbx;
	open my $log_handle, '>', \$log;

	$converter->input_fh( $mrc_handle );
	$converter->tbx_fh($tbx_handle);
	$converter->log_fh($log_handle);
	$converter->convert;

	close $mrc_handle;
	close $tbx_handle;
	close $log_handle;

	#remove datetime stamps from log
	$log = remove_datetime(undef, $log);

	#return output TBX and log
	return [$tbx, $log];
}

sub no_tbx {
	my ($tbx_log) = @_;

	return [ undef, $tbx_log->[1] ];
}

sub remove_datetime {
	my ($text) = @_;
	defined $text
		or return;
	$text =~ s/\] \[[^\]]+\]/\]/gm;
	return $text;
}

# fix version numbers and chomp logs
# so that they can be matched properly
sub fix_version {
	my ($log) = @_;

	chomp $log;
	#fix version
	$log =~ s/\[version\]/Convert::MRC->_version/e;

	return $log;
}

1;


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "CONVERT MRC TO TBX-BASIC",
  "dist_author" => [
    "Nathan Rasmussen, Nathan Glenn <garfieldnate\@gmail.com>"
  ],
  "dist_name" => "Convert-MRC",
  "dist_version" => "0.01",
  "license" => "perl",
  "module_name" => "Convert::MRC",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "English" => 0,
    "Exporter" => 0,
    "Log::Message::Simple" => 0,
    "base" => 0,
    "open" => 0,
    "perl" => "5.008",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [
    "bin/mrc2tbx"
  ],
  "test_requires" => {
    "Data::Section::Simple" => 0,
    "File::Find" => 0,
    "File::Slurp" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "File::Touch" => 0,
    "FindBin" => 0,
    "Test::Base" => 0,
    "Test::Base::Filter" => 0,
    "Test::LongString" => 0,
    "Test::More" => "0.96",
    "Test::XML" => 0,
    "autodie" => 0
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Convert/MRC/Variables.pm
package Convert::MRC::Variables;

use strict;
use warnings;
# ABSTRACT: Provide global constants used by Convert::MRC
# VERSION
use base 'Exporter';
## no critic (ProhibitAutomaticExportation)
our @EXPORT = qw(
  %corresp
  $langCode
  %correctCaps
  %allowed
  %legalIn
  %position
  %meta
);

# reference variables
our %corresp     = _get_corresp();
our $langCode    = _get_lang_code();
our %correctCaps = _get_correct_caps();
our %allowed     = _get_allowed();
our %legalIn     = _get_legal_in();
our %position    = _get_position();
our %meta        = _get_meta();

# How does the data category from a header row relate to the header?
# (This is also a validity check.)
sub _get_corresp {
    my %corresp = (
        workingLanguage => 'Language',
        sourceDesc      => 'Source',
        subjectField    => 'Subject',
    );
    return %corresp;
}

# ISO 639 language code, and optionally region code: fr, eng-US
# case-insensitive; values are smashed to lowercase when parsed
sub _get_lang_code {
    return qr/[a-zA-Z]{2,3}(?:-[a-zA-Z]{2})?/;
}

# What is the proper capitalization for each data category/picklist item?
# A hash from a case-smashed version to the correct version, which will be
# used to recognize and fix user input.
sub _get_correct_caps {
    my %correct_caps;
    $correct_caps{'DatCat'}{ lc($_) } = $_ foreach qw (
      sourceDesc workingLanguage subjectField xGraphic definition
      term partOfSpeech administrativeStatus context geographicalUsage
      grammaticalGender termLocation termType note source
      crossReference externalCrossReference customerSubset projectSubset
      transactionType fn org title role email uid tel adr type
    );
    $correct_caps{'partOfSpeech'}{ lc($_) } = $_ foreach qw (
      noun verb adjective adverb properNoun other
    );
    $correct_caps{'administrativeStatus'}{ lc($_) } = $_ foreach qw (
      preferredTerm-admn-sts admittedTerm-admn-sts
      deprecatedTerm-admn-sts supersededTerm-admn-sts
    );
    $correct_caps{'grammaticalGender'}{ lc($_) } = $_ foreach qw (
      masculine feminine neuter other
    );
    $correct_caps{'termLocation'}{ lc($_) } = $_ foreach qw (
      menuItem dialogBox groupBox textBox comboBox comboBoxElement
      checkBox tab pushButton radioButton spinBox progressBar slider
      informativeMessage interactiveMessage toolTip tableText
      userDefinedType
    );
    $correct_caps{'termType'}{ lc($_) } = $_ foreach qw (
      fullForm acronym abbreviation shortForm variant phrase
    );
    $correct_caps{'transactionType'}{ lc($_) } = $_ foreach qw (
      origination modification
    );
    $correct_caps{'type'}{ lc($_) } = $_ foreach qw (
      person organization
    );
    return %correct_caps;
}

# Which additional fields are allowed on which data categories?
sub _get_allowed {
    my %allowed;
    $allowed{$_}{'Note'}   = 1 foreach qw ();
    $allowed{$_}{'Source'} = 1 foreach qw (
      definition subjectField context
    );
    $allowed{$_}{'Link'} = 1 foreach qw (
      transactionType crossReference externalCrossReference xGraphic
    );
    $allowed{'transactionType'}{'Date'} =
      $allowed{'transactionType'}{'Responsibility'} = 1;
    $allowed{$_}{'FieldLang'} = 1 foreach qw (Source Note Responsibility);
    return %allowed;
}

# which data categories are allowed at which level?
sub _get_legal_in {
    $legalIn{'Concept'}{$_} = 1 foreach qw (
      transactionType crossReference externalCrossReference
      customerSubset projectSubset xGraphic subjectField note
      source
    );
    $legalIn{'LangSet'}{$_} = 1 foreach qw (transactionType crossReference
      externalCrossReference customerSubset projectSubset
      definition note source
    );
    $legalIn{'Term'}{$_} = 1
      foreach qw (transactionType crossReference externalCrossReference
      customerSubset projectSubset context grammaticalGender
      geographicalUsage partOfSpeech termLocation termType
      administrativeStatus note source term
    );
    $legalIn{'Party'}{$_} = 1 foreach qw (email title role org uid tel adr fn);
    return %legalIn;
}

# what part of the term structure does each data category go in?
sub _get_position {
    my %position = map { $_ => 'termGrp' } qw (
      administrativeStatus geographicalUsage grammaticalGender
      partOfSpeech termLocation termType
    );
    %position = (
        %position,
        map { $_ => 'auxInfo' }
          qw (
          context customerSubset projectSubset crossReference note
          source transactionType externalCrossReference xGraphic
          )
    );
    return %position;
}

# which TBX meta data category does each data category print as?
sub _get_meta {
    my %meta;
    $meta{$_} = 'admin'   foreach qw (customerSubset projectSubset);
    $meta{$_} = 'descrip' foreach qw (definition subjectField context);
    $meta{$_} = 'termNote'
      foreach
      qw (grammaticalGender geographicalUsage partOfSpeech termLocation termType administrativeStatus);
    return %meta;
}

1;


Makefile.PL

use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "CONVERT MRC TO TBX-BASIC",
  "AUTHOR" => "Nathan Rasmussen, Nathan Glenn <garfieldnate\@gmail.com>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Convert-MRC",
  "EXE_FILES" => [
    "bin/mrc2tbx"
  ],
  "LICENSE" => "perl",
  "NAME" => "Convert::MRC",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "English" => 0,
    "Exporter" => 0,
    "Log::Message::Simple" => 0,
    "base" => 0,
    "open" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Section::Simple" => 0,
    "File::Find" => 0,
    "File::Slurp" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "File::Touch" => 0,
    "FindBin" => 0,
    "Test::Base" => 0,
    "Test::Base::Filter" => 0,
    "Test::LongString" => 0,
    "Test::More" => "0.96",
    "Test::XML" => 0,
    "autodie" => 0
  },
  "VERSION" => "0.01",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



